### ISSUE TITLE
JPA-Map storage might loose writes due to missing locking mechanism

### ISSUE THREAD

**Issue by ahus1** [keycloak/keycloak#9411]
### Describe the bug

The JPA implementation for the map storage provider doesn't specify a locking strategy, therefore writes might get lost: When Hibernate writes the contents back to the database, the contents might have been updated by another concurrent request already.
In an extreme case this could lead to some column being updated while others aren't, leaving the database in an inconsistent state. 

### Version

17.0.0-SNAPSHOT

### Expected behavior

When persisting data in the JPA map storage provider 
succeeds, the contents should be stored consistently in all related tables (main entity table plus additional tables for searching). 

When a concurrent transaction modifies the data, the transaction should fail with an exception. 

### Actual behavior

When a concurrent write occurred, Hibernate will overwrite them without noticing some other transaction has modified the data in between. As Hibernate will write only content to the database it considered changed, this could lead to a partial overwrite of the concurrent transaction, leading to inconsistent data.

### How to Reproduce?

Looking at the code, no JPA entity of the map storage has a version column. The option of `optimistic-lock="all"` is not used AFAIK.

While the old JPA storage uses PESSIMISTIC_WRITE in several places (usually when deleting data), there is no such code in the new JPA storage provider.

### Anything else?

**Inside of the scope of this issue:** The old storage uses PESSIMISTIC_WRITE in several places, find out the rationale why it did so, and if this should apply to the new JPA map storage as well.

**Outside of the scope of this issue:** As the admin API of Keycloak doesn't provide a mechanism for optimistic locking or identification of concurrent updates by multiple clients (there is not version ID or similar in the API), lost writes when accessing the REST API might still occur. Adding a mechanism for that in the API of map storage and other APIs is out of scope here. This is only about keeping the JPA storage consistent.

**Comment by ahus1** [comment]
One part of a possible solution would be to add a `@Version` column to the entities in the map-jpa storage. 

How and when pessimistic write should be used still needs to be investigated.

**Comment by ahus1** [comment]
The results I described above would depend on the isolation level set of the JDBC connection. I see that the JpaMapStorageProviderFactory issues a warning on the console when a isolation level of less than REPEATABLE_READ is specified. 

The default for a non-configured isolation level, the code states "READ_COMMITTED". 

Instead of a `@Version` column to support "READ_COMMITTED", it would also suffice to enforce an isolation level of REPEATABLE_READ, plus handling phantom reads in one or the other way, for example by explicitly locking the existing rows, see https://vladmihalcea.com/phantom-read/

Higher isolation levels usually come with reduced DB performance AFAIK, still I won't be able to tell how much of a difference it would make. 

I looked at the existing Keycloak documentation, and didn't find a user-configurable setting on recommended settings for isolation levels.
